src/main/scala/org/apache/spark/shuffle/ucx/UcxWorkerWrapper.scala

```diff
@@ -0,0 +1,151 @@
+                       memoryPool: MemoryPool)
+  extends Closeable with Logging {
+
+  // To keep connection map on a remote side by id rather then by worker address, which could be big.
+  // Would not need when migrate to active messages.
+  private val id: Array[Byte] = (transport.executorId + s"_${Thread.currentThread().getId}").getBytes
+  private final val connections = mutable.Map.empty[String, UcpEndpoint]
+  private var transactionId = Random.nextLong()
+  private val workerAddress = worker.getAddress
+
+  override def close(): Unit = {
+    connections.foreach{
+      case (_, endpoint) => endpoint.close()
+    }
+    connections.clear()
+    worker.close()
+  }
+
+  /**
+   * The only place for worker progress
+   */
+  private[ucx] def progress() {
+    if ((worker.progress() == 0) && ucxConf.useWakeup) {
+      worker.waitForEvents()
+      worker.progress()
+    }
+  }
+
+  private def getConnection(executorId: String): UcpEndpoint = {
+    val workerAdresses = transport.executorIdToAddress
+
+    if (!workerAdresses.contains(executorId)) {
+      // Block until there's no worker address for this BlockManagerID
+      val startTime = System.currentTimeMillis()
+      val timeout = ucxConf.conf.getTimeAsMs("spark.network.timeout", "100")
+      workerAdresses.synchronized {
+        while (workerAdresses.get(executorId) == null) {
+          workerAdresses.wait(timeout)
+          if (System.currentTimeMillis() - startTime > timeout) {
+            throw new UcxException(s"Didn't get worker address for $executorId during $timeout")
+          }
+        }
+      }
+    }
+
+    connections.getOrElseUpdate(executorId, {
+      logInfo(s"Worker from thread ${Thread.currentThread().getName} connecting to $executorId")
+      val endpointParams = new UcpEndpointParams()
+        .setUcpAddress(workerAdresses.get(executorId))
+      worker.newEndpoint(endpointParams)
+    })
+  }
+
+  def fetchBlockByBlockId(executorId: String, blockId: BlockId,
+                          resultBuffer: MemoryBlock, cb: OperationCallback): UcxRequest = {
+    val stats = new UcxStats()
+    val ep = getConnection(executorId)
+    val mem = memoryPool.get(transport.ucxShuffleConf.rpcMessageSize)
+    val buffer = UcxUtils.getByteBufferView(mem.address,
+      transport.ucxShuffleConf.rpcMessageSize.toInt)
+
+    transactionId += 1
+    val tag = transactionId
+    buffer.putLong(tag)
+    buffer.putInt(id.length)
+    buffer.put(id)
+    workerAddress.rewind()
+    buffer.putInt(workerAddress.capacity())
+    buffer.put(workerAddress)
+
+    Utils.tryWithResource(new ByteBufferBackedOutputStream(buffer)) { bos =>
+      val out = new ObjectOutputStream(bos)
+      out.writeObject(blockId)
+      out.flush()
+    }
+
+    logTrace(s"Sending message to $executorId to fetch $blockId on tag $tag," +
+      s"resultBuffer $resultBuffer")
+    ep.sendTaggedNonBlocking(mem.address, transport.ucxShuffleConf.rpcMessageSize, tag,
```

Comment:
so this tag is from the random number? If I understand correctly, this is sending a message to someone else, using a random tag.  How's the tag being being received?

Reply:
Ok I think I see, the receiver doesn't care and the tag used on the next send, is the one you sent her in line 121.